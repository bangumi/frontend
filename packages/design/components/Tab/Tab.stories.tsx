import React from 'react'
import Tab from '.'

export default {
  title: 'modern/Tabs',
  component: Tab,
  argTypes: {
    activeKey: {
      options: ['all', 'animation', 'book', 'music', 'game', 'drama'],
      control: { type: 'select' }
    },
    items: {
      control: { type: 'object' }
    }
  }
}

const Template = ({ activeKey }: {activeKey: string}) => {
  const items = [{
    key: 'all',
    label: '全部条目'
  }, {
    key: 'animation',
    label: '动画'
  }, {
    key: 'book',
    label: '书籍'
  }, {
    key: 'music',
    label: '音乐'
  }, {
    key: 'game',
    label: '游戏'
  }, {
    key: 'drama',
    label: '三次元'
  }]
  return (
    <Tab items={items} activeKey={activeKey}>
    </Tab>
  )
}

export const Usage = Template.bind({})
