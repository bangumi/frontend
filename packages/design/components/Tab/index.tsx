import React, { PropsWithChildren } from 'react'
import classnames from 'classnames'

interface ItemProps {
  isActive: boolean
  type: 'default' | 'borderless'
  onClick: () => void
}
const Item: React.FC<PropsWithChildren<ItemProps>> = ({ children, isActive, onClick, type }) => {
  return (
    <li
      className={classnames('bgm-tab__item', `bgm-tab__item--${type}`, { 'bgm-tab__item--active': isActive })}
      onClick={onClick}
    >
      {children}
    </li>
  )
}

interface ItemType {
  key: string
  label: string
}

export interface TabProps {
  /* 节点数组 */
  items: ItemType[]
  /* 选中节点的 Key */
  activeKey: string
  /* 点击切换事件回调，对每一个 Item 都生效 */
  onChange?: (key: string) => void
  /* 样式类型 */
  type?: 'default' | 'borderless'
}

export const Tab: React.FC<TabProps> = ({ activeKey, items, onChange, type = 'default' }) => {
  return (
    <ul className={classnames('bgm-tab', `bgm-tab--${type}`)}>
      {
      items.map((item) => {
        return (
          <Item
            key={item.key}
            isActive={activeKey === item.key}
            type={type}
            onClick={() => { onChange?.(item.key) }}
          > {item.label}
          </Item>
        )
      })
    }
    </ul>
  )
}

export default Tab
