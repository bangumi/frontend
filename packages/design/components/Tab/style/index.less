.tab {
  display: inline-flex;
  flex-direction: row;

  padding: 4px;

  list-style: none;
  background-color: #FACDD0;
  border-radius: 19px;

  // TODO: use normalize?
  margin-block-start: 0;
  margin-block-end: 0;
  margin-inline-start: 0;
}

.tab-item {
  font-size: 14px;
  font-weight: bold;
  line-height: 14px;
  padding: 4px 6px;
  font-family: 'Source Han Sans SC';
  color: #FFFFFF;
  cursor: pointer;

  &.active {
    background-color: #F09199;
    border-radius: 18px;
  }
}