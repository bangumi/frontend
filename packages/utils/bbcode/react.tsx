import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
