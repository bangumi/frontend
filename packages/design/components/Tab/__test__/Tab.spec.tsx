import { render } from '@testing-library/react';
import React from 'react';

import Tab from '..';

it('should render all tab', () => {
  const items = [
    {
      key: '1',
      label: '1',
    },
    {
      key: '2',
      label: '2',
    },
  ];
  const { getByText } = render(<Tab items={items} activeKey='1' />);
  expect(getByText('1')).toBeInTheDocument();
  expect(getByText('2')).toBeInTheDocument();
  expect(getByText('1')).toHaveClass('bgm-tab__item--active');
});

it('should trigger onChange when click tabs', () => {
  const items = [
    {
      key: '1',
      label: '1',
    },
    {
      key: '2',
      label: '2',
    },
  ];
  const handleChange = vi.fn();
  const { getByText } = render(<Tab items={items} activeKey='1' onChange={handleChange} />);
  getByText('2').click();

  expect(handleChange).toBeCalledWith('2', items[1]);
});

it('should render borderless tab', () => {
  const items = [
    {
      key: '1',
      label: '1',
    },
    {
      key: '2',
      label: '2',
    },
  ];
  const { container } = render(<Tab items={items} activeKey='1' type='borderless' />);
  expect(container).toMatchSnapshot();
});
