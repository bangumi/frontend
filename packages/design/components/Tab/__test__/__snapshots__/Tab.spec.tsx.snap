// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render borderless tab 1`] = `
<div>
  <ul
    class="bgm-tab bgm-tab--borderless"
  >
    <li
      class="bgm-tab__item bgm-tab__item--borderless bgm-tab__item--active"
    >
      1
    </li>
    <li
      class="bgm-tab__item bgm-tab__item--borderless"
    >
      2
    </li>
  </ul>
</div>
`;
